### DISCUSSION TITLE
How to send HOSTNAME to Router?

### DISCUSSION THREAD

**Discussion by Mike0675** [cesanta/mongoose#2558] [Ask questions, get answers] [answered]
Another dumb Question:
I´using PJRC´s Teensy4.1 what is working well now.

Now I´m looking for a way to announce the HOSTNAME of my Device to the Router.
I´m using static IP.

On ESP32 this is done when connecting to WiFi. 

But how can it be done with build-in TCP-Stack?

Thanks for Answers

**Comment by scaprile** [comment]
> I´m looking for a way to announce the HOSTNAME of my Device to the Router.

Do you know what that means and the options available ? What _exactly_ do you mean by that ? Routers do not "know" hostnames (nor hosts, btw), routers know networks. What is "the router" to you ? What _exactly_ do you want to do ?

That is not "done by ESP32", it doesn't do magic nor abstract stuff to  please you; what _exactly_ do you want to do ? Do you have a DNS ? Do you know how DNS works ? mDNS ? NetBIOS ? [SSDP](https://mongoose.ws/documentation/tutorials/udp-ssdp-search/) ? Port forwarding ? I have a limited imagination, I may have left one or two (hundred) options out.

**Reply by Mike0675** [reply]
Hi scaprile, 
Thank you for your patience.

I´m using a "All in One" WiFi/DSL/Network-Router Named "Fritzbox7490".
My Wish is to access my Teensy4.1 in the Browser with its Hostname like so: http://Hostname
The ESP32 sends this Hostname when it get his IP over DHCP to the Router.
This works; but (as I testet right now) not when I set a static IP.
I think thats because DHCP isn´t used.

Now back to Teensy:
Whats the best way to do it? 
Which Protocols are supported by mongoose?
Is SSDP a Option? So  I will try this.
Is mDNS supported?
Where to set Hostname for DHCP?

**Comment by scaprile** [comment, marked as answer]
To do things properly, you should have a DNS server.
Some DNS and DHCP servers integrate together. Your router probably does that. I don't know if relying on user-given host names is standard, considered good practice, or even allowed. Sounds insecure to me, but looks fine for a (closed) home network
Mongoose does not have mDNS nor SSDP support, but no one is stopping you from broadcasting UDP datagrams.
In Windows environments, some people run a netbios name server (or however it is called these days) and Windows machines used to resolve with those queries. Not a standard, but used to work
We use a bogus hostname, are you able to reach your board at http://mip ?
https://github.com/cesanta/mongoose/blob/b8bfa5565b10000b38a10780ad0af72babbb662d/src/net_builtin.c#L267-L281
You can change that.

**Reply by Mike0675** [reply]
Thanks you, thats what I´m looked for.

Yes, with DHCP mongoose sends mip to the Router amd can be reached under http://mip.

I´m don´t know if this is unsecure, but Espressifs ESP is doin so.

Can I change mongoose.c like that?

```
uint8_t opts[] = { 
       53, 1, 3,                 // Type: DHCP request 
       55, 2, 1,   3,            // GW and mask 
       12, 3, 'h', 'o', 's', 't', 'n', 'a', 'm', 'e'     // Host name: "mip" 
       54, 4, 0,   0,   0,   0,  // DHCP server ID 
       50, 4, 0,   0,   0,   0,  // Requested IP 
       255                       // End of options 
   }; 
```

This solution is not my favorite ,because I have to change Code in mongoose.c.

Is there another possibility to change that?

Thanks a lot!

**Comment by scaprile** [comment]
That one is a short hardcoded hack of a function, it isn't that easy to change. You'll see there are some hardcoded offsets there that you need to take care of if you change the text. And... '3' will no longer be '3' before your host name... take a look at the DHCP request message structure.
We currently have other priorities as none of our customers have requested that feature, but I'll ask if we can have it configurable somehow. It is usually the MAC address that is used to assign an IP address, not the announced host name.
